### ISSUE TITLE
Lab 6 Level 3

### ISSUE THREAD

**Issue by alvintjw** [nus-cs2030/2223-s2#466]
# Summary

Hi Can someone explain this level? Not quite sure what my KeyableMap is suppose to entail. Is it suppose to have a string and immutable map just like student and module? I understand that we are suppose to have student and module extend from keyable map so are we suppose to override both the get and put method in student and module? What would the return type be for the KeyableMap get and put method then?

## Description:

Clarifications or details with respect to the issue you are facing

## Code Snippets

Attach code snippets (if any)

## Screenshots

Screenshots (if any)

**Comment by Huanqing-Mao** [comment]
you are not supposed to override get method, as you should use generics to define the get(...), but we need to override the put method in each child class so that the return type is correct

**Comment by territxl** [comment]
KeyableMap is like a class the generically define the functions such as put() which can be generalised for the other classes like Module. However as mentioned by Huanqing, do ensure your return type is correct for the different classes. Although the put() method is similar for those classes, the return type is different.

**Comment by alvintjw** [comment]
@territxl In module and student, I have an attribute called map that's of type ImmutableMap and in the get function(), I am calling this.map.get(key).  So am I suppose to define the same for keyable to have an attribute ImmutableMap and using generics to be able to store an immutablmap of <String,assessments> and <string,module>?

**Comment by littlemidget3** [comment]
yeah leave keyable as it is. KeyableMap should encompass the put, get, and other shared methods between Student and module. Student and module will then extend from KeyableMap. Depending on your implementation from level 1 and 2, you may need to do some tweaks. Be precise in in your method and class definitions, as it may cause a lot of problems

**Comment by jefflcy** [comment]
hey alvin, it is also important that you type parameterize correctly for the methods in the KeyableMap class! :)

**Comment by alvintjw** [comment]
@jefflcy thanks jepp:)
